[Freshness] US-1953530: Refresh top articles by PV - .NET and Languages Content (5) (#30303)
dotnet/docs: docs/standard/base-types/best-practices.md

[!code-csharp[Conceptual.RegularExpressions.BestPractices#4](../../../samples/snippets/csharp/VS_Snippets_CLR/conceptual.regularexpressions.bestpractices/cs/static2.cs#4)]
[!code-vb[Conceptual.RegularExpressions.BestPractices#4](../../../samples/snippets/visualbasic/VS_Snippets_CLR/conceptual.regularexpressions.bestpractices/vb/static2.vb#4)]

By default, the last 15 most recently used static regular expression patterns are cached. For applications that require a larger number of cached static regular expressions, the size of the cache can be adjusted by setting the <xref:System.Text.RegularExpressions.Regex.CacheSize%2A?displayProperty=nameWithType> property.

The regular expression `\p{Sc}+\s*\d+` that's used in this example verifies that the input string has a currency symbol and at least one decimal digit. The pattern is defined as shown in the following table:

|Pattern|Description|
|-------------|-----------------|
|`\p{Sc}+`|Matches one or more characters in the Unicode Symbol, Currency category.|
|`\s*`|Matches zero or more white-space characters.|
|`\d+`|Matches one or more decimal digits.|

### Interpreted vs. compiled regular expressions

> Regular expression patterns that aren't bound to the regular expression engine through the specification of the <xref:System.Text.RegularExpressions.RegexOptions.Compiled> option are interpreted. When a regular expression object is instantiated, the regular expression engine converts the regular expression to a set of operation codes. When an instance method is called, the operation codes are converted to MSIL and executed by the JIT compiler. Similarly, when a static regular expression method is called and the regular expression can't be found in the cache, the regular expression engine converts the regular expression to a set of operation codes and stores them in the cache. It then converts these operation codes to MSIL so that the JIT compiler can execute them. Interpreted regular expressions reduce startup time at the cost of slower execution time. Because of this process, they're best used when the regular expression is used in a few number of method calls, or if the exact number of calls to regular expression methods is unknown but is expected to be less. As the number of method calls increases, the performance gain from reduced startup time is outstripped by the slower execution speed.

Reviewer: ```suggestion
Regular expression patterns that aren't bound to the regular expression engine through the specification of the <xref:System.Text.RegularExpressions.RegexOptions.Compiled> option are interpreted. When a regular expression object is instantiated, the regular expression engine converts the regular expression to a set of operation codes. When an instance method is called, the operation codes are converted to MSIL and executed by the JIT compiler. Similarly, when a static regular expression method is called and the regular expression can't be found in the cache, the regular expression engine converts the regular expression to a set of operation codes and stores them in the cache. It then converts these operation codes to MSIL so that the JIT compiler can execute them. Interpreted regular expressions reduce startup time at the cost of slower execution time. Because of this process, they're best used when the regular expression is used in a small number of method calls, or if the exact number of calls to regular expression methods is unknown but is expected to be small. As the number of method calls increases, the performance gain from reduced startup time is outstripped by the slower execution speed.
```

To summarize, we recommend that you use interpreted regular expressions when you call regular expression methods with a specific regular expression relatively infrequently. You should use compiled regular expressions when you call regular expression methods with a specific regular expression relatively frequently. It's difficult to determine the exact threshold at which the slower execution speeds of interpreted regular expressions outweigh gains from their reduced startup time, or the threshold at which the slower startup times of compiled regular expressions outweigh gains from their faster execution speeds. It depends on various factors, including the complexity of the regular expression and the specific data that it processes. To determine whether interpreted or compiled regular expressions offer the best performance for your particular application scenario, you can use the <xref:System.Diagnostics.Stopwatch> class to compare their execution times.

> The following example compares the performance of compiled and interpreted regular expressions when reading the first 10 sentences and when reading all the sentences in the text of Theodore Dreiser's *The Financier*. As the output from the example shows, when only 10 calls are made to regular expression matching methods, an interpreted regular expression offers better performance than a compiled regular expression. However, a compiled regular expression offers better performance when a large number of calls (in this case, over 13,000) is made.

Reviewer: ```suggestion
The following example compares the performance of compiled and interpreted regular expressions when reading the first 10 sentences and when reading all the sentences in the text of Theodore Dreiser's *The Financier*. As the output from the example shows, when only 10 calls are made to regular expression matching methods, an interpreted regular expression offers better performance than a compiled regular expression. However, a compiled regular expression offers better performance when a large number of calls (in this case, over 13,000) are made.
```

Reviewer: It doesn't sound right to me with "is" and also based on [this](http://grammartips.homestead.com/number.html#:~:text=When%20deciding%20whether%20to%20use%20a%20singular%20or,ARE%20The%20Number%20Is%20%3B%20A%20Number%20Are), I think "are" is correct because the calls are acting separately.

[!code-csharp[Conceptual.RegularExpressions.BestPractices#5](../../../samples/snippets/csharp/VS_Snippets_CLR/conceptual.regularexpressions.bestpractices/cs/compare1.cs#5)]
[!code-vb[Conceptual.RegularExpressions.BestPractices#5](../../../samples/snippets/visualbasic/VS_Snippets_CLR/conceptual.regularexpressions.bestpractices/vb/compare1.vb#5)]

The regular expression pattern used in the example, `\b(\w+((\r?\n)|,?\s))*\w+[.?:;!]`, is defined as shown in the following table:

|Pattern|Description|
|-------------|-----------------|

> |`\b`|Begins the match at a word boundary.|

Reviewer: ```suggestion
|`\b`|Begin the match at a word boundary.|
```

Reviewer: If you don't want it to be a command, then it should be something like "Specifies that the match is to begin at a word boundary".

|`\w+`|Matches one or more word characters.|
|<code>(\r?\n)&#124;,?\s)</code>|Matches either zero or one carriage return followed by a newline character, or zero or one comma followed by a white-space character.|

> |<code>(\w+((\r?\n)&#124;,?\s))*</code>|Matches zero or more occurrences of one or more word characters. These characters are followed either by zero or one carriage return and a newline character, or by zero or one comma followed by a white-space character.|

Reviewer: ```suggestion
|<code>(\w+((\r?\n)&#124;,?\s))*</code>|Matches zero or more occurrences of one or more word characters that are followed either by zero or one carriage return and a newline character, or by zero or one comma followed by a white-space character.|
```

|`\w+`|Matches one or more word characters.|
|`[.?:;!]`|Matches a period, question mark, colon, semicolon, or exclamation point.|

### Regular expressions: Compiled to an assembly

> .NET also enables you to create an assembly that contains compiled regular expressions. This characteristic moves the performance hit of regular expression compilation from run time to design time. However, it also involves some additional work. You must define the regular expressions in advance and compile them to an assembly. The compiler can then reference this assembly when compiling source code that uses the assembly's regular expressions. Each compiled regular expression in the assembly is represented by a class that derives from <xref:System.Text.RegularExpressions.Regex>.

Reviewer: ```suggestion
.NET also enables you to create an assembly that contains compiled regular expressions. This capability moves the performance hit of regular expression compilation from run time to design time. However, it also involves some additional work. You must define the regular expressions in advance and compile them to an assembly. The compiler can then reference this assembly when compiling source code that uses the assembly's regular expressions. Each compiled regular expression in the assembly is represented by a class that derives from <xref:System.Text.RegularExpressions.Regex>.
```

> To compile regular expressions to an assembly, you call the <xref:System.Text.RegularExpressions.Regex.CompileToAssembly%28System.Text.RegularExpressions.RegexCompilationInfo%5B%5D%2CSystem.Reflection.AssemblyName%29?displayProperty=nameWithType> method and pass it an array of <xref:System.Text.RegularExpressions.RegexCompilationInfo> objects. The objects represent the regular expressions to be compiled, and an <xref:System.Reflection.AssemblyName> object that contains information about the assembly to be created.

Reviewer: ```suggestion
To compile regular expressions to an assembly, you call the <xref:System.Text.RegularExpressions.Regex.CompileToAssembly%28System.Text.RegularExpressions.RegexCompilationInfo%5B%5D%2CSystem.Reflection.AssemblyName%29?displayProperty=nameWithType> method and pass it an array of <xref:System.Text.RegularExpressions.RegexCompilationInfo> objects and an <xref:System.Reflection.AssemblyName> object. The <xref:System.Text.RegularExpressions.RegexCompilationInfo> objects represent the regular expressions to be compiled, and the <xref:System.Reflection.AssemblyName> object contains information about the assembly to be created.
```

We recommend that you compile regular expressions to an assembly in the following situations:

- If you're a component developer who wants to create a library of reusable regular expressions.

> - If you expect your regular expression's pattern-matching methods to be called an indeterminate number of times anywhere from once or twice to thousands or tens of thousands of times. Unlike compiled or interpreted regular expressions, regular expressions that are compiled to separate assemblies offer performance that's consistent regardless of the number of method calls.

Reviewer: ```suggestion
- If you expect your regular expression's pattern-matching methods to be called an indeterminate number of times&mdash;anywhere from once or twice to thousands or tens of thousands of times. Unlike compiled or interpreted regular expressions, regular expressions that are compiled to separate assemblies offer performance that's consistent regardless of the number of method calls.
```

> If you're using compiled regular expressions to optimize performance, you shouldn't use reflection to create the assembly, load the regular expression engine, and execute its pattern-matching methods. This process requires that you avoid building regular expression patterns dynamically, and that you specify any pattern-matching options, such as case-insensitive pattern matching at the time the assembly is created. It also requires that you separate the code that creates the assembly from the code that uses the regular expression.

Reviewer: ```suggestion
If you're using compiled regular expressions to optimize performance, you shouldn't use reflection to create the assembly, load the regular expression engine, and execute its pattern-matching methods. Avoiding reflection requires that you don't build regular expression patterns dynamically, and that you specify any pattern-matching options, such as case-insensitive pattern matching, at the time the assembly is created. It also requires that you separate the code that creates the assembly from the code that uses the regular expression.
```

Support for backtracking gives regular expressions power and flexibility. It also places the responsibility for controlling the operation of the regular expression engine in the hands of regular expression developers. Because developers are often not aware of this responsibility, their misuse of backtracking or reliance on excessive backtracking often plays the most significant role in degrading regular expression performance. In a worst-case scenario, execution time can double for each additional character in the input string. In fact, by using backtracking excessively, it's easy to create the programmatic equivalent of an endless loop if input nearly matches the regular expression pattern. The regular expression engine might take hours or even days to process a relatively short input string.

> Often, applications pay a performance penalty for using backtracking although backtracking isn't essential for a match. For example, the regular expression `\b\p{Lu}\w*\b` matches all words that begin with an uppercase character, as the following table shows:

Reviewer: ```suggestion
Often, applications pay a performance penalty for using backtracking even though backtracking isn't essential for a match. For example, the regular expression `\b\p{Lu}\w*\b` matches all words that begin with an uppercase character, as the following table shows:
```

|Pattern|Description|
|-|-|

> |`\b`|Begins the match at a word boundary.|

Reviewer: ```suggestion
|`\b`|Begin the match at a word boundary.|
```

|`\p{Lu}`|Matches an uppercase character.|
|`\w*`|Matches zero or more word characters.|

> |`\b`|Ends the match at a word boundary.|

Reviewer: ```suggestion
|`\b`|End the match at a word boundary.|
```

If you determine that backtracking isn't necessary, you can disable it by using the `(?>subexpression)` language element, known as an atomic group. The following example parses an input string by using two regular expressions. The first, `\b\p{Lu}\w*\b`, relies on backtracking. The second, `\b\p{Lu}(?>\w*)\b`, disables backtracking. As the output from the example shows, they both produce the same result:

[!code-csharp[Conceptual.RegularExpressions.BestPractices#10](../../../samples/snippets/csharp/VS_Snippets_CLR/conceptual.regularexpressions.bestpractices/cs/backtrack2.cs#10)]
[!code-vb[Conceptual.RegularExpressions.BestPractices#10](../../../samples/snippets/visualbasic/VS_Snippets_CLR/conceptual.regularexpressions.bestpractices/vb/backtrack2.vb#10)]

In many cases, backtracking is essential for matching a regular expression pattern to input text. However, excessive backtracking can severely degrade performance and create the impression that an application has stopped responding. In particular, this problem arises when quantifiers are nested and the text that matches the outer subexpression is a subset of the text that matches the inner subexpression.

> [!WARNING]

> > In addition to avoiding excessive backtracking, you must use the timeout feature to ensure that excessive backtracking doesn't severely degrade regular expression performance. For more information, see the [Use Time-out Values](#use-time-out-values) section.

Reviewer: ```suggestion
> In addition to avoiding excessive backtracking, you should use the timeout feature to ensure that excessive backtracking doesn't severely degrade regular expression performance. For more information, see the [Use Time-out Values](#use-time-out-values) section.
```

For example, the regular expression pattern `^[0-9A-Z]([-.\w]*[0-9A-Z])*\$$` is intended to match a part number that consists of at least one alphanumeric character. Any additional characters can consist of an alphanumeric character, a hyphen, an underscore, or a period, though the last character must be alphanumeric. A dollar sign terminates the part number. In some cases, this regular expression pattern can exhibit poor performance because quantifiers are nested, and because the subexpression `[0-9A-Z]` is a subset of the subexpression `[-.\w]*`.

In these cases, you can optimize regular expression performance by removing the nested quantifiers and replacing the outer subexpression with a zero-width lookahead or lookbehind assertion. Lookahead and lookbehind assertions are anchors. They don't move the pointer in the input string but instead look ahead or behind to check whether a specified condition is met. For example, the part number regular expression can be rewritten as `^[0-9A-Z][-.\w]*(?<=[0-9A-Z])\$$`. This regular expression pattern is defined as shown in the following table:

|Pattern|Description|
|-------------|-----------------|

> |`^`|Begins the match at the beginning of the input string.|

Reviewer: ```suggestion
|`^`|Begin the match at the beginning of the input string.|
```

|`[0-9A-Z]`|Matches an alphanumeric character. The part number must consist of at least this character.|
|`[-.\w]*`|Matches zero or more occurrences of any word character, hyphen, or period.|
|`\$`|Matches a dollar sign.|

> |`(?<=[0-9A-Z])`|Looks ahead of the ending dollar sign to ensure that the previous character is alphanumeric.|

Reviewer: ```suggestion
|`(?<=[0-9A-Z])`|Look ahead of the ending dollar sign to ensure that the previous character is alphanumeric.|
```

> |`$`|Ends the match at the end of the input string.|

Reviewer: ```suggestion
|`$`|End the match at the end of the input string.|
```

|Language element|Description|
|----------------------|-----------------|
|`(?=` `subexpression` `)`|Zero-width positive lookahead. Looks ahead of the current position to determine whether `subexpression` matches the input string.|
|`(?!` `subexpression` `)`|Zero-width negative lookahead. Looks ahead of the current position to determine whether `subexpression` doesn't match the input string.|
|`(?<=` `subexpression` `)`|Zero-width positive lookbehind. Looks behind the current position to determine whether `subexpression` matches the input string.|
|`(?<!` `subexpression` `)`|Zero-width negative lookbehind. Looks behind the current position to determine whether `subexpression` doesn't match the input string.|

## Use time-out values

> If your regular expressions processes input that nearly matches the regular expression pattern, it can often rely on excessive backtracking, which impacts its performance significantly. In addition to carefully considering your use of backtracking and testing the regular expression against near-matching input, you must always set a time-out value to minimize the effect of excessive backtracking, if it occurs.

Reviewer: ```suggestion
If your regular expressions processes input that nearly matches the regular expression pattern, it can often rely on excessive backtracking, which impacts its performance significantly. In addition to carefully considering your use of backtracking and testing the regular expression against near-matching input, you should always set a time-out value to minimize the effect of excessive backtracking, if it occurs.
```

The regular expression time-out interval defines the period of time that the regular expression engine will look for a single match before it times out. Depending on the regular expression pattern and the input text, the execution time might exceed the specified time-out interval, but it won't spend more time backtracking than the specified time-out interval. The default time-out interval is <xref:System.Text.RegularExpressions.Regex.InfiniteMatchTimeout?displayProperty=nameWithType>, which means that the regular expression won't time out. You can override this value and define a time-out interval as follows:

- Call the <xref:System.Text.RegularExpressions.Regex.%23ctor%28System.String%2CSystem.Text.RegularExpressions.RegexOptions%2CSystem.TimeSpan%29> constructor to provide a time-out value when you instantiate a <xref:System.Text.RegularExpressions.Regex> object.

> - Call a static pattern matching method, such as <xref:System.Text.RegularExpressions.Regex.Match%28System.String%2CSystem.String%2CSystem.Text.RegularExpressions.RegexOptions%2CSystem.TimeSpan%29?displayProperty=nameWithType> or <xref:System.Text.RegularExpressions.Regex.Replace%28System.String%2CSystem.String%2CSystem.String%2CSystem.Text.RegularExpressions.RegexOptions%2CSystem.TimeSpan%29?displayProperty=nameWithType> that includes a `matchTimeout` parameter.

Reviewer: ```suggestion
- Call a static pattern matching method, such as <xref:System.Text.RegularExpressions.Regex.Match%28System.String%2CSystem.String%2CSystem.Text.RegularExpressions.RegexOptions%2CSystem.TimeSpan%29?displayProperty=nameWithType> or <xref:System.Text.RegularExpressions.Regex.Replace%28System.String%2CSystem.String%2CSystem.String%2CSystem.Text.RegularExpressions.RegexOptions%2CSystem.TimeSpan%29?displayProperty=nameWithType>, that includes a `matchTimeout` parameter.
```

- Call the constructor that has a parameter of type <xref:System.TimeSpan> for compiled regular expressions that are created by calling the <xref:System.Text.RegularExpressions.Regex.CompileToAssembly%2A?displayProperty=nameWithType> method.

> - Set a process-wide or AppDomain-wide value with code, such as `AppDomain.CurrentDomain.SetData("REGEX_DEFAULT_MATCH_TIMEOUT", TimeSpan.FromMilliseconds(100));`.

Reviewer: ```suggestion
- Set a process-wide or AppDomain-wide value with code such as `AppDomain.CurrentDomain.SetData("REGEX_DEFAULT_MATCH_TIMEOUT", TimeSpan.FromMilliseconds(100));`.
```

Reviewer: In this case the clause is restrictive so there shouldn't be a comma before 'such as'. See https://www.grammarly.com/blog/such-as-comma/.

The following example defines a `GetWordData` method that instantiates a regular expression with a time-out interval of 350 milliseconds to calculate the number of words and average number of characters in a word in a text document. If the matching operation times out, the time-out interval is increased by 350 milliseconds and the <xref:System.Text.RegularExpressions.Regex> object is reinstantiated. If the new time-out interval exceeds one second, the method rethrows the exception to the caller.

[!code-csharp[Conceptual.RegularExpressions.BestPractices#12](../../../samples/snippets/csharp/VS_Snippets_CLR/conceptual.regularexpressions.bestpractices/cs/timeout1.cs#12)]
[!code-vb[Conceptual.RegularExpressions.BestPractices#12](../../../samples/snippets/visualbasic/VS_Snippets_CLR/conceptual.regularexpressions.bestpractices/vb/timeout1.vb#12)]

## Capture only when necessary

> Regular expressions in .NET support many grouping constructs, which let you group a regular expression pattern into one or more subexpressions. The most commonly used grouping constructs in .NET regular expression language are `(`*subexpression*`)`, which defines a numbered capturing group, and `(?<`*name*`>`*subexpression*`)`, which defines a named capturing group. Grouping constructs are essential for creating backreferences and for defining a subexpression to which a quantifier is applied.

Reviewer: ```suggestion
Regular expressions in .NET support grouping constructs, which let you group a regular expression pattern into one or more subexpressions. The most commonly used grouping constructs in .NET regular expression language are `(`*subexpression*`)`, which defines a numbered capturing group, and `(?<`*name*`>`*subexpression*`)`, which defines a named capturing group. Grouping constructs are essential for creating backreferences and for defining a subexpression to which a quantifier is applied.
```

Often, grouping constructs are used in a regular expression only so that quantifiers can be applied to them. The groups captured by these subexpressions aren't used later. For example, the regular expression `\b(\w+[;,]?\s?)+[.?!]` is designed to capture an entire sentence. The following table describes the language elements in this regular expression pattern and their effect on the <xref:System.Text.RegularExpressions.Match> object's <xref:System.Text.RegularExpressions.Match.Groups%2A?displayProperty=nameWithType> and <xref:System.Text.RegularExpressions.Group.Captures%2A?displayProperty=nameWithType> collections:

|Pattern|Description|
|-------------|-----------------|

> |`\b`|Begins the match at a word boundary.|

Reviewer: ```suggestion
|`\b`|Begin the match at a word boundary.|
```
